### PR TITLE
🌱 ClusterClass: add condition for references with outdated apiVersions 

### LINKED ISSUE #7212: Improve ClusterAPI upgrade workflow with ClusterClass
**User Story**

As a user I would like to have a smooth Cluster API upgrade process with ClusterClass-based clusters.

### Current state

*Note*: The following is only relevant if the Cluster API upgrade contains apiVersion bumps of control plane / bootstrap / infra provider CRDs.

ClusterClass inline patches are coupled to a specific template apiVersion. However, the Cluster topology controller always uses the latest apiVersion during reconciliation. When a template apiVersion is bumped the pre-existing inline patches won't be applied anymore.

In the following example the ClusterClass has been written for CAPD v1beta1. If now the CAPD CRDs are bumped to v1beta2 the patches won't apply anymore as the topology controller is using v1beta2 automatically. This will trigger a rollout of the Cluster without the patches.
```yaml
kind: ClusterClass
[...]
infrastructure:
  ref:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: DockerMachineTemplate
    name: quick-start-docker-worker-machinetemplate
   namespace: default
- apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
  kind: DockerMachineTemplate
  matchResources:
    controlPlane: true
---
kind: CustomResourceDefinition
metadata:
  labels:
    cluster.x-k8s.io/provider: infrastructure-docker
    cluster.x-k8s.io/v1beta1: v1beta1_v1beta2
  name: dockermachinetemplates.infrastructure.cluster.x-k8s.io
```

The only workaround as of today is to pause all Clusters before the Cluster API upgrade, then upgrade all ClusterClasses to the new apiVersion and then unpause all Clusters. The goal of this issue is to decouple the ClusterAPI upgrade from upgrading the ClusterClass to use the new apiVersion.

### Proposal: Cluster topology controller should use apiVersions from the ClusterClass

The topology controller would use the exact apiVersions specified in the ClusterClass and not the latest available.

The update workflow then looks like this: (example with CAPD v1beta1=>v1beta2)
* ClusterAPI and ClusterClass-based clusters are deployed
* ClusterAPI is upgraded
* Cluster topology reconciliation still works
  * The reconciler still uses CAPD v1beta1 objects to reconcile the cluster
* ClusterClass is upgraded to use CAPD v1beta2 types
  * The template refs and the patches are upgraded at the same time to v1beta2
  * This step doesn't have to be done directly after the upgrade, but eventually before v1beta1 is dropped

Tasks:
* [ ] Additionally we should drop the automatic apiVersion bump in the ClusterClass controller. As of today the bump doesn't work anyway as the request to bump the apiVersion in the template ref fails, because then the selectors don't match anymore.
* [ ] Adjust Cluster topology controller to read templates from ClusterClass and current objects from the managed topology via the version specified in the ClusterClass
* [ ] Let's consider adding a condition to the ClusterClass which surfaces if a ClusterClass is using not the latest apiVersions.

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7212

### CODE REVIEW

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 1]
```diff
@@ -133,37 +132,74 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
...
-		if err := r.reconcileExternal(ctx, clusterClass, ref, !patchedRefs.Has(uniqueKey)); err != nil {
+
+		// Continue as we only have to reconcile every referenced object once.
+		if reconciledRefs.Has(uniqueKey) {
+			continue
+		}
+
+		reconciledRefs.Insert(uniqueKey)
+
+		if err := r.reconcileExternal(ctx, clusterClass, ref); err != nil {
```
Comment (fabriziopandini) [inline internal/controllers/clusterclass/clusterclass_controller.go:154]:
```suggestion
		// Add the ClusterClass as owner reference to the template
		// so clusterctl move can identify all the related objects.
		if err := r.reconcileExternal(ctx, clusterClass, ref); err != nil {
```
note: I'm starting to think that we can bake in clusterctl move some specific knowledge of our types, but this is a different discussion

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 2]
```diff
@@ -133,37 +132,74 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
...
 			errs = append(errs, err)
 			continue
 		}
-		patchedRefs.Insert(uniqueKey)
+
+		updatedRef := ref.DeepCopy()
```
Comment (fabriziopandini) [inline internal/controllers/clusterclass/clusterclass_controller.go:159]:
```suggestion
		// Check if the template reference is outdated (it is not using the latest API version for the current CAPI contract)
		updatedRef := ref.DeepCopy()
```

**Review by fabriziopandini — COMMENTED** [review]
First pass

WRt to question above:
- I'm +1 to use Using a more specific condition like APIVersionsUpToDate, or simply ReferenceUpToDate (I prefer the second one)
- I'm -1 to use Ready condition because it is reserved for being used as a summary of all conditions; however, I will not add ready at the current stage (it is optional in the policy)

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:154]: Interesting idea. Should be way more efficient then setting ownerRefs everywhere (but I guess in a lot of cases we want them for delete use-cases anyway)

Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:154]: Mentioned both
```go
		// Add the ClusterClass as owner reference to the templates so clusterctl move 
		// can identify all related objects and Kubernetes garbage collector deletes 
		// all referenced templates on ClusterClass deletion.
```

**Comment by sbueringer** [comment]
> First pass
> 
> WRt to question above:
> 
> * I'm +1 to use Using a more specific condition like APIVersionsUpToDate, or simply ReferenceUpToDate (I prefer the second one)
> * I'm -1 to use Ready condition because it is reserved for being used as a summary of all conditions; however, I will not add ready at the current stage (it is optional in the policy)

Thx good point, adjusted accordingly

**Comment by sbueringer** [comment]
Should be ready for another round of review now

File: api/v1beta1/condition_consts.go [hunk 3]
```diff
@@ -282,3 +282,16 @@ const (
 	// not yet completed because at least one of the lifecycle hooks is blocking.
 	TopologyReconciledHookBlockingReason = "LifecycleHookBlocking"
 )
+
+// Conditions and condition reasons for ClusterClass.
+const (
+	// ClusterClassReferencesUpToDateCondition documents if the references in the ClusterClass are
+	// up-to-date (i.e. they are using the latest apiVersion of the current Cluster API contract from
+	// the corresponding CRD).
+	ClusterClassReferencesUpToDateCondition ConditionType = "ReferencesUpToDate"
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/condition_consts.go:291]: I prefer `RefVersionsUpToDate`

File: api/v1beta1/condition_consts.go [hunk 4]
```diff
@@ -282,3 +282,16 @@ const (
...
+
+	// ClusterClassReferencesNotUpToDateReason (Severity=Warning) that the references in the ClusterClass are not
+	// up-to-date (i.e. they are using the latest apiVersion of the current Cluster API contract from
+	// the corresponding CRD).
+	ClusterClassReferencesNotUpToDateReason = "ReferencesNotUpToDate"
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/condition_consts.go:296]: Fine for me. I'll defer to Fabrizio

**Comment by sbueringer** [comment]
@fabriziopandini When you have some time, would be good to get your input on condition / reason names

File: api/v1beta1/condition_consts.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply api/v1beta1/condition_consts.go:291]: +1 to RefVersionsUpToDate

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
/assign @vincepri 
for final approval

File: api/v1beta1/condition_consts.go [hunk 5]
```diff
@@ -282,3 +282,16 @@ const (
...
+// Conditions and condition reasons for ClusterClass.
+const (
+	// ClusterClassRefVersionsUpToDateCondition documents if the references in the ClusterClass are
+	// up-to-date (i.e. they are using the latest apiVersion of the current Cluster API contract from
+	// the corresponding CRD).
+	ClusterClassRefVersionsUpToDateCondition ConditionType = "RefVersionsUpToDate"
+
+	// ClusterClassOutdatedRefVersionsReason (Severity=Warning) that the references in the ClusterClass are not
+	// up-to-date (i.e. they are using the latest apiVersion of the current Cluster API contract from
+	// the corresponding CRD).
```
Comment (ykakarap) [inline api/v1beta1/condition_consts.go:295]:
```suggestion
	// up-to-date (i.e. they are not using the latest apiVersion of the current Cluster API contract from
	// the corresponding CRD).
```
I see it working both ways. Feel free to resovle if you prefer the current wording.

File: internal/controllers/clusterclass/clusterclass_controller.go [hunk 6]
```diff
@@ -133,14 +149,57 @@ func (r *Reconciler) reconcile(ctx context.Context, clusterClass *clusterv1.Clus
...
+		for currentRef, updatedRef := range outdatedRefs {
+			msg = append(msg, fmt.Sprintf("%s should be %s", currentRef.GroupVersionKind().String(), updatedRef.GroupVersionKind().String()))
+		}
+		conditions.Set(
+			clusterClass,
+			conditions.FalseCondition(
+				clusterv1.ClusterClassRefVersionsUpToDateCondition,
+				clusterv1.ClusterClassOutdatedRefVersionsReason,
+				clusterv1.ConditionSeverityWarning,
+				strings.Join(msg, ", "),
```
Comment (ykakarap) [inline internal/controllers/clusterclass/clusterclass_controller.go:193]: How about using a different seperator like `;` (or something else). GVK when converted to string already has `,` in it. The output string could be confusing.


Also, the message only uses the GVK of the ref which does not include the name. If a few DockerMachineTemplate refs are outdated that tt could be confusing if the mesage just repeats that a DockerMachineTemplate ref is outdated without pointing to the correct ref.

File: api/v1beta1/condition_consts.go (hunk 5 above)
Comment (sbueringer) [reply api/v1beta1/condition_consts.go:295]: You're absolutely right, fixed

File: internal/controllers/clusterclass/clusterclass_controller.go (hunk 6 above)
Comment (sbueringer) [reply internal/controllers/clusterclass/clusterclass_controller.go:193]: I changed it to use a quoted string. Keep using `,` as separater seems slight more grammatically correct. The condition now looks like this:
```yaml
  status:
    conditions:
    - lastTransitionTime: "2022-10-07T10:04:22Z"
      message: Ref "infrastructure.cluster.x-k8s.io/v1alpha4, Kind=DockerClusterTemplate
        default/quick-start-my-cluster" should be "infrastructure.cluster.x-k8s.io/v1beta1,
        Kind=DockerClusterTemplate default/quick-start-my-cluster", Ref "infrastructure.cluster.x-k8s.io/v1alpha4,
        Kind=DockerMachineTemplate default/quick-start-control-plane" should be "infrastructure.cluster.x-k8s.io/v1beta1,
        Kind=DockerMachineTemplate default/quick-start-control-plane", Ref "infrastructure.cluster.x-k8s.io/v1alpha4,
        Kind=DockerMachineTemplate default/quick-start-docker-worker-machinetemplate"
        should be "infrastructure.cluster.x-k8s.io/v1beta1, Kind=DockerMachineTemplate
        default/quick-start-docker-worker-machinetemplate"
      reason: OutdatedRefVersions
      severity: Warning
      status: "False"
      type: RefVersionsUpToDate
```
It's a bit verbose, but also very explicit which is probably a good thing

**Pull request merged**